webapp/src/controllers/apikey.ts

```diff
@@ -0,0 +1,181 @@
+	return res.json({ ...data, account: res.locals.account });
+}
+
+export async function addKeyApi(req, res, next) {
+	const { name, description, expirationDays, ownerId } = req.body;
+
+	let validationError = chainValidations(
+		req.body,
+		[
+			{ field: 'name', validation: { notEmpty: true, ofType: 'string' } },
+			{ field: 'expirationDays', validation: { notEmpty: true, ofType: 'string' } },
+			{ field: 'ownerId', validation: { notEmpty: true } }
+		],
+		{
+			name: 'Name',
+			expirationDays: 'Number of days before expiry',
+			ownerId: 'Id of the creator of the key'
+		}
+	);
+	if (validationError) {
+		return dynamicResponse(req, res, 400, { error: validationError });
+	}
+
+	//generate expiration date
+	const epxirationDate = new Date();
+	if (expirationDays === 'never') {
+		epxirationDate.setDate(epxirationDate.getDate() + 365); //need a better way to set never
+	} else {
+		const numOfDays = parseInt(expirationDays);
+		epxirationDate.setDate(epxirationDate.getDate() + numOfDays);
+	}
+
+	//add key to db
+	const addedKey = await addKey({
+		name: name,
+		description: description,
+		expirationDate: epxirationDate,
+		ownerId: toObjectId(ownerId)
+	});
+
+	//generate key for the user
+	const keyToken = jwt.sign(
+		{ keyId: addedKey?.insertedId, accountId: ownerId, version: 0 },
+		process.env.JWT_SECRET,
+		expirationDays == 'never' ? {} : { expiresIn: `${expirationDays}d` }
+	);
+
+	console.log('keyToken', keyToken);
+	//add the token back into the key object
+	const updatedKey = await addTokenToKey(addedKey?.insertedId, keyToken);
+
+	return dynamicResponse(req, res, 302, {
+		keyId: addedKey.insertedId,
+		ownerId: ownerId,
+		token: keyToken,
+		redirect: '/apikeys'
+	});
+}
+
+export async function deleteKeyApi(req, res, next) {
+	const { keyId, ownerId } = req.body;
+
+	let validationError = chainValidations(
+		req.body,
+		[
+			{ field: 'keyId', validation: { notEmpty: true } },
+			{ field: 'ownerId', validation: { notEmpty: true } }
+		],
+		{ keyId: 'Key', ownerId: 'Key Creator' }
+	);
+	if (validationError) {
+		return dynamicResponse(req, res, 400, { error: validationError });
+	}
+
+	await deleteKey(toObjectId(ownerId), toObjectId(keyId));
+
+	return dynamicResponse(req, res, 302, {
+		//redirect: `/apikeys`
+	});
+}
+
+export async function incrementKeyApi(req, res, next) {
+	const { keyId, ownerId } = req.body;
+
+	let validationError = chainValidations(
+		req.body,
+		[{ field: 'keyId', validation: { notEmpty: true } }],
+		{ keyId: 'Key' }
+	);
+	if (validationError) {
+		return dynamicResponse(req, res, 400, { error: validationError });
+	}
+
+	await incrementVersion(toObjectId(ownerId), toObjectId(keyId));
+
+	const modifiedKey = await getKeyById(toObjectId(ownerId), toObjectId(keyId));
+
+	const newToken = jwt.sign(
+		{ keyId: modifiedKey?._id, ownerId: modifiedKey?.ownerId, version: modifiedKey?.version },
+		process.env.JWT_SECRET
+	);
+
+	await addTokenToKey(toObjectId(modifiedKey?._id), newToken);
+
```

Comment:
Calculate date difference between current date and the original key expiry, and add the `expiresIn` option to the new signed key with the number of seconds remaining